operators/src/concurrency/thread_pool.rs

```diff
@@ -0,0 +1,339 @@
+        );
+
+        let worker_deques: Vec<Worker<Task>> =
+            (0..number_of_threads).map(|_| Worker::new_fifo()).collect();
+
+        let mut thread_pool = Self {
+            degree_of_parallelism: number_of_threads,
+            global_queue: Arc::new(Injector::new()),
+            stealers: worker_deques.iter().map(Worker::stealer).collect(),
+            threads: Vec::with_capacity(number_of_threads),
+            next_group_id: AtomicCell::new(0),
+        };
+
+        for worker_deque in worker_deques {
+            let global_queue = thread_pool.global_queue.clone();
+            let stealers = thread_pool.stealers.clone();
+
+            thread_pool.threads.push(std::thread::spawn(move || {
+                Self::await_work(&worker_deque, &global_queue, &stealers);
+            }))
+        }
+
+        thread_pool
+    }
+
+    fn await_work(local: &Worker<Task>, global: &Injector<Task>, stealers: &[Stealer<Task>]) {
+        let backoff = Backoff::new();
+
+        loop {
+            // Pop a task from the local queue, if not empty.
+            let task = local.pop().or_else(|| {
+                // Otherwise, we need to look for a task elsewhere.
+                iter::repeat_with(|| {
+                    // Try stealing a batch of tasks from the global queue.
+                    global
+                        .steal_batch_and_pop(local)
+                        // Or try stealing a task from one of the other threads.
+                        .or_else(|| stealers.iter().map(Stealer::steal).collect())
+                })
+                // Loop while no task was stolen and any steal operation needs to be retried.
+                .find(|s| !s.is_retry())
+                // Extract the stolen task, if there is one.
+                .and_then(Steal::success)
+            });
+
+            if let Some(task) = task {
+                // TODO: recover panics
+                task.call_once();
+                backoff.reset();
+            } else {
+                // TODO: sleep instead of busy waiting?
+                backoff.snooze();
+            }
+        }
+    }
+
+    pub fn create_context(&self) -> ThreadPoolContext {
+        ThreadPoolContext::new(self, self.next_group_id.fetch_add(1))
+    }
+
+    fn compute(&self, task: Task) {
+        self.global_queue.push(task);
+    }
+
+    // TODO: scope for threads
+}
+
+/// A computation context for a group that spawns tasks in a `ThreadPool`
+#[derive(Copy, Clone, Debug)]
+pub struct ThreadPoolContext<'pool> {
+    thread_pool: &'pool ThreadPool,
+    task_group_id: TaskGroupId,
+}
+
+impl<'pool> ThreadPoolContext<'pool> {
+    /// Create a new `ThreadPoolContext`
+    fn new(thread_pool: &'pool ThreadPool, task_group_id: TaskGroupId) -> Self {
+        Self {
+            thread_pool,
+            task_group_id,
+        }
+    }
+
+    /// What is the degree of paralellism that the `ThreadPool` provides?
```

Comment:
- typo parallelism
- I don't get this measurement, what is the unit? can't we just stick to number of threads?

Reply:
No, if we for instance decide later on, that we spawn more tasks if some get stuck on tasks for too long, then we still want to provide the operators who use it a degree of parallelism to decide on how to split the work.

E.g., our goal is to provide 1 thread per core, so 8 threads, but for reasons we spawn some more in the pool, we still don't want to create more than 8 tasks for computing items for an array.

Reply:
So degree of parallelism would be the target number of threads but the actual number of threads may vary if we spawn additional ones because there is no progress in some tasks? Maybe call it `target_thread_count` and `current_thread_count` then, because `degree` is not really measurable here. We'd also need some parameter how many additional threads may be spawned but we can do it later.

Reply:
ef3bfce 

I renamed the variable but not the function. The variable name was misleading, but the function reflects this: https://en.wikipedia.org/wiki/Degree_of_parallelism